Ex jdbc01

verbusharas opened the pull request:
Labas Elvinai,
Prašau, gelbėk!
Kadangi vis gaudavau pastabų kad negalima kartoti kodo, tai bandžiau padaryti kuo abstraktesnį variantą kaip bendrauti su MySQL duombaze per JDBC.
Nežinau kiek valandų jau praleidau bandydamas įvairius variantus bet vis kažkas nesigauna. 

Ar tikrai įmanoma išvengti kodo kartojimų, kai persistinam primityviu JDBC būdu? vistiek juk statement'ai kartojasi tik kad keičiasi Stringai skliausteliuose priklausomai nuo objekto sudarymo principų...

Na kaip bebūtų, siunčiu kodą, kurį išmąsčiau abstractinant ir kuris galų gale vistiek neveikia... 
Gal užteks kad nukreipsi kokianors reikiama kryptimi?
Bandysiu gal dar per kokius Dao factory ar kažkaip...

verbusharas committed:
cleanup

verbusharas committed:
JDBC excercise 1 solved. Working, but can be improved (e.g. about the person id retrieval process)

verbusharas committed:
improved person appointment to project process (surname instead of id) - still needs improving since surnames can be equal

verbusharas committed:
improved person appointment to project process (surname instead of id) - still needs improving since surnames can be equal

verbusharas committed:
attempted to implement SOLID principles and to make everything abstract. Failed miserably. Will need to retry with DAOs.

verbusharas committed:
attempted to implement SOLID principles and to make everything abstract. Failed miserably. Will need to retry with DAOs.

verbusharas committed:
attempted to implement SOLID principles and to make everything abstract. Failed miserably. Will need to retry with DAOs.

verbusharas committed:
Removed .idea directory

verbusharas committed:
Removed out directory

verbusharas committed:
attempted to implement SOLID principles and to make everything abstract. Failed miserably. Will need to retry with DAOs.

verbusharas committed:
remove file

verbusharas committed:
remove target directory

Elvinasl commented on Ex_JDBC01/src/main/java/lt/verbus/config/MySqlConfig.java:
```diff
@@ -0,0 +1,9 @@
+
+    public static final String SQL_URL = "jdbc:mysql://localhost:3306/employees";
+    public static final String SQL_USERNAME = "root";
+    public static final String SQL_PASSWORD = "08princeAli";
```
Keistai atrodo public static password :) As loadinciau is .properties failo

verbusharas replied:
Sutapimo lygis: prieš kelias minutes Udemy kurse būtent apie properties išmokau. Bet čia Springe per @Value

Elvinasl commented on Ex_JDBC01/src/main/java/lt/verbus/model/Project.java:
```diff
@@ -0,0 +1,45 @@
+    public Project(Integer id, String title) {
+        this.id = id;
+        this.title = title;
+        this.executorList = new ArrayList<>();
```
As deklaruociau tiesiog  private List<Executor> executorList = new ArrayList<>();

Tavo variantas veiks, bet atrodo keistai. :)

Elvinasl commented on Ex_JDBC01/src/main/java/lt/verbus/services/CrudService.java:
```diff
@@ -0,0 +1,16 @@
+
+    void delete(T t);
+
+}
```
Atrodo sauniai 👍

Elvinasl commented on Ex_JDBC01/src/main/java/lt/verbus/services/DatabaseService.java:
```diff
@@ -0,0 +1,21 @@
+    List<Project> findAllProjects() throws SQLException;
+    void addNewPerson(Executor executor) throws SQLException;
+    void appointPersonToProject(Executor executor, Project project) throws SQLException;
+
```
Realiai Person ir Project kiekvienas atskirai turi savo servisa. Pvz ProjectService, PersonService ir pan. Cia ir yra monolitine architektura: https://microservices.io/patterns/monolithic.html 

Tu jau turi CrudService, ja gali panaudoti cia ir perpanaudot jos metodus.

Elvinasl commented on Ex_JDBC01/src/main/java/lt/verbus/services/DatabaseService.java:
```diff
@@ -0,0 +1,21 @@
+
+    // Iš pažiūros klasė "tightly coupled" tarp Project ir Person tipų
+    // Ar verta daryti Generic interface? T.y. DatabaseService<T>
+    // bet tada pasimes logika, kaip priskirinėti person to project... ?
```
Jeigu tau reikia rysio Person -> Project (arba atvirksciai), tu gali injectinti tau reikalinga servisa. Manau galbut jums dar per anktsti, bet pasiskaityk apie dependency injection (spring supportina).

https://www.baeldung.com/spring-dependency-injection#:~:text=Dependency%20Injection%20is%20a%20fundamental,managing%20components%20onto%20the%20container.

Elvinasl commented on Ex_JDBC01/src/main/java/lt/verbus/services/TableConverter.java:
```diff
@@ -0,0 +1,101 @@
+//        }
+//
+//
+//
```
Kodel git'e yra palikta komentaru? Tam gitas ir yra, jeigu noresim pasiziuret kaip kodas anksciau atrode, mes galesim nueit i senesni commit'a. Issivalykim.

Elvinasl commented on Ex_JDBC01/src/main/java/lt/verbus/services/TableConverter.java:
```diff
@@ -0,0 +1,101 @@
+        List<TargetType> targetedList = new ArrayList<>();
+        for (Convertable convertable : convertTableToGenericList())
+            targetedList.add((TargetType)convertable);
+        return targetedList;
```
Noreciau cia matyti Stream :)

Elvinasl commented on Ex_JDBC01/src/main/java/lt/verbus/services/TableConverter.java:
```diff
@@ -0,0 +1,101 @@
+            objectSet.add(new ConvertableImpl(id, name));
+    }
+        return new ArrayList<>(objectSet);
+    }
```
Idomus metodas, sukuri set, tada while loop'as, tada konvertuojam i Array? Keista.

Gal gali ir cia panaudoti stream api?

Elvinasl commented on Ex_JDBC01/src/main/java/lt/verbus/AppJDBC01.java:
```diff
@@ -0,0 +1,40 @@
+//
+//        System.out.println(databaseTemp.getAllProjects());
+//
+//        databaseTemp.closeConnections();
```
As realiai tikeciausi tokios strukturos:

Services
 - PersonService
 - ProjectService
 - N + 1 servisu

Repository
 - PersonRepository (implements CrudRepository)
 - ProjectRepository (implements CrudRepository)
 - CrudRepository (kuri turi add, update, get, delete)

Model
 - Project
 - Person
ir t.t.

Config
- DbConfigo reikalingos klases

verbusharas replied:
Jeigu PersonRepository, ProjectRepository atlikinėja CRUD operacijas, tuomet kokia paskirtis Servisų? 
Taip pat - iš kur turėtų pareiti duombazės connection instance? iš Config package?

Elvinasl replied:
Repositorijos bus atsakingos uz custom querius, pvz GetPersonsByName(String name); O crud repositorija bus bendra, kur tures common operacijas.

Servisuose bus tavo biznio logika, gi ne visada reikia tik duomenu bazes uzklausos, kartais reikia ir apdoroti duomenis. Tam ir yra skirtas servisas.

Elvinasl replied:
Apie duomenu baze turetu tureti savo klase, as konfiguruociau config package.

verbusharas replied:
Ačiū Elvinai. Dabar jau, manau, aišku :)

Elvinasl commented on Ex_JDBC01/src/main/java/lt/verbus/services/DatabaseServiceImpl.java:
```diff
@@ -0,0 +1,61 @@
+//        preparedStatement.setInt(2, executor.getId());
+//        preparedStatement.executeUpdate();
+//        preparedStatement.close();
+    }
```
Manau person ir project logika turetu buti isskirta i atskiras klases (PersonService/ProjectService)

verbusharas committed:
Restructured the project to have appropriate Service and Repository hierarchy.
Haven't yet figured out how to provide SQL connection properties from .properties file.

verbusharas commented:
@Elvinasl please review

Elvinasl commented on Ex_JDBC01/src/main/java/lt/verbus/repository/CrudRepository.java:
```diff
@@ -0,0 +1,20 @@
+
+public interface CrudRepository<ID, T extends BaseEntity> {
+
+    Map<ID, T> findAll() throws SQLException;
```
O kodel cia map'as? As tikeciausi kolekcijos

Elvinasl replied:
Cia noreciau matyt Collection<T>. Tada kvieciant sita funkcija, galesi pasiimt ar List<T> arba Set<T> ir pan.

Elvinasl commented on Ex_JDBC01/src/main/java/lt/verbus/repository/ProjectRepository.java:
```diff
@@ -0,0 +1,91 @@
+            if (projectExecutorsInRepo.get(executor.getId()) == null) {
+                appointExecutorToProject(executor, project);
+            }
+        }
```
Cia noreciau kad panaudotum Stream

Elvinasl commented on Ex_JDBC01/src/main/java/lt/verbus/model/BaseEntity.java:
```diff
@@ -0,0 +1,15 @@
+    public void setId(Integer id) {
+        this.id = id;
+    }
+
```
Kazkaip silpnai, kam kurti atskira klase, jeigu tik ID turi? Gal tada idek bent koki createdAt, updatedAt, isActive? Logiskiau butu.

Elvinasl commented on Ex_JDBC01/src/main/java/lt/verbus/repository/ExecutorRepository.java:
```diff
@@ -0,0 +1,87 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public class ExecutorRepository implements CrudRepository<Integer, Executor> {
```
Speju cia bandai viena 'globalu' updateri? Bet matai, tu vistiek esi prisirises prie enticio, ir juos esi hardcodines (pvz "table.getString("qualification")"). Gal galim padaryti kazka universalaus? Pasiziurek i tai:

Class<?> clazz
clazz.getDeclaredFields() <-- Sitas tau grazintu kolekcija fieldu, is kuriu galetum pasiimti fieldName (pvz qualification)

verbusharas committed:
Moved configs to properties file, used streams forEach instead of regular forEach
